src/libraries/System.Private.CoreLib/src/System/Text/TranscodingStream.cs

```diff
@@ -0,0 +1,598 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Text
+{
+    internal sealed class TranscodingStream : Stream
+    {
+        private const int DefaultReadByteBufferSize = 4 * 1024; // lifted from StreamReader.cs (FileStream)
+
+        // We optimistically assume 1 byte ~ 1 char during transcoding. This is a good rule of thumb
+        // but isn't always appropriate: transcoding between single-byte and multi-byte encodings
+        // will violate this, as will any invalid data fixups performed by the transcoder itself.
+        // To account for these unknowns we have a minimum scratch buffer size we use during the
+        // transcoding process. This should be generous enough to account for even the largest
+        // fallback mechanism we're likely to see in the real world.
+
+        private const int MinWriteRentedArraySize = 4 * 1024;
+        private const int MaxWriteRentedArraySize = 1024 * 1024;
+
+        private readonly Encoding _innerEncoding;
+        private readonly Encoding _thisEncoding;
+        private Stream? _innerStream; // null if the wrapper has been disposed
```

Comment:
Since use after Dispose is considered a contract violation, and nullable reference types are all about a contract, we've only been considering the nullable annotations to be valid until Dispose.  So this can be non-nullable.  Then in Dispose when you null it out, just use `null!`.  That also the removes some of the warnings you had to suppress.